### ISSUE TITLE
Maxwell Running Out of Memory

### ISSUE THREAD

**Issue by hongd4356** [zendesk/maxwell#816]
We are trying to use maxwell as a way to stream data from a mysql server to kinesis and eventually elastic search. We have a large amount of data in mysql that we need to bootstrap, and it appears that we are running out of memory when we attempt too many bootstraps or bootstraps with massive amounts of data. Based on the CPU usage, it looks like the KPL refuses to release memory after the bootstrap process is finished which eventually leads to a memory crash.  To be clear, we are bootstrapping by inserting into the bootstrap table and not using the bootstrap utility, and we are using the maxwell running process. 

We have a few questions:
Have you seen this issue before?
Is this related to maxwell processes or is this an issue with the KCL?
Is there a setting in the Kinesis properties that can mitigate this issue?

**Comment by osheroff** [comment]
@thomasdziedzic any input here?

@hongd4356 when you say "after the bootstrap process is finished", do you mean that all the data has successfully hit kinesis?  Or is data still streaming to kinesis?

**Comment by thomasdziedzic** [comment]
sorry, i don't know

**Comment by gautamvij** [comment]
@osheroff 
Facing same issue, In my case the data has successfully hit kinesis and successfully processed at other end.

**Comment by ede-n** [comment]
@osheroff Yes after successfully processing. We had a where clause for different ranges of ids and corresponding records in `bootstrap` table. The first row's `bootstrap.is_complete` flag is set to `1` and the process died. No errors on the Kinesis end were recorded.

**Comment by gautamvij** [comment]
@osheroff 
Since, KPL stores records in memory as FutureCallback. In case of large records it pushes all records in memory leading to memory spike. 

We can throttle ingestion in kinesis stream by waiting to let backpressure be processed as mentioned in this aws [blog](https://aws.amazon.com/blogs/big-data/implementing-efficient-and-reliable-producers-with-the-amazon-kinesis-producer-library/).

@osheroff Let me know if this is acceptable, will give the PR.

**Comment by osheroff** [comment]
backpressure mechanism merged and released.
